controller/readwrite.go

```diff
@@ -76,10 +76,14 @@ func (rw *ReadWrite) Run(ctx context.Context) error {
 		rw.watcherCh = make(chan string, rw.drivers.Len()+2)
 	}
 	go func() {
-		err := rw.watcher.Watch(ctx, rw.watcherCh)
-		if err != nil {
-			rw.logger.Error("error watching template dependencies", "error", err)
-			errCh <- err
+		for {
+			rw.logger.Trace("starting template dependency monitoring")
+			err := rw.watcher.Watch(ctx, rw.watcherCh)
+			if err == nil || err == context.Canceled {
```

Comment:
Just to make sure I understand, `rw.watcher.Watch()` would, in the typical case, not return unless there's an error because it's constantly watching the dependencies. If there is a non context canceled error, we want to instead of retry watching instead of returning. But if there is a case where we do want to stop watching the dependency or we're going through a shutdown where the context was canceled, then we exit. Is that accurate?

Reply:
Yep! the idea is for it to not return in the typical case.

For the retry: Retry handling is [set by CTS](https://github.com/hashicorp/consul-terraform-sync/blob/main/controller/watcher.go#L70) and handled by hcat on a per-dependency case. This PR/main-as-is will have CTS retry until it maxes out the retry attempts ~15mins, and then the dependency would probably sit idle(?) and the tasks no longer trigger on it. I think the ideal case here would be for us to move towards high availability is to have indefinite retries with a max wait duration (instead of exponential).

Shut down where the context is canceled: yep!

Stop watching dependency: kind of, the context portion here would be the main program shutting down, and then would relay to the watcher to stop, stopping all dependencies ([source](https://github.com/hashicorp/consul-terraform-sync/blob/main/controller/controller.go#L76-L78)).

Stop watching individual dependencies: CTS task -> Terraform template -> multiple hcat dependencies; and different templates can have the same dependencies as another template. If we want to stop monitoring a subset of dependencies, CTS will need to remove the task's template from hcat, and then hcat uses that to determine if there are any other templates relying on it. If none, only then would hcat stop them, otherwise they will keep watching. I think it's the [`watcher.Sweep` method](https://pkg.go.dev/github.com/hashicorp/hcat#Watcher.Sweep)? I haven't looked into this too carefully yet since the delete task is a new feature.

Hope this helps! Let me know if there are other thoughts around this -- more questions or ideas

Reply:
That was very helpful, thanks for explaining with such detail!